src/aws-lambda/alexa-skill-lambda/alexa-skill-lambda.py

```diff
@@ -0,0 +1,1414 @@
+        point1: Array-like with 2 members (lat/long).
+        point2: Array-like with 2 members (lat/long).
+
+    Returns:
+        float: distance in kilometres
+    """
+    return ((point1[0] - point2[0]) ** 2 + (point1[1] - point2[1]) ** 2) ** 0.5 * 111
+
+
+def get_customer_location():
+    """
+    Get customer faked location (we could also read it from device via Alexa
+    but we might be demo-ing from the web where location
+    is not available - see
+    https://developer.amazon.com/en-US/docs/alexa/custom-skills/location-services-for-alexa-skills.html
+    ).
+    Returns:
+        list: Coordinates lat/long.
+    """
+    cstore_route = json.loads(urlopen(f'{LOCATION_SERVICE_URL}/cstore_route').read().decode('utf-8'))
+    customer_position = cstore_route['features'][0]['geometry']['coordinates'][0]
+    return customer_position
+
+
+def location_search_cstore() -> Tuple[str, float]:
+    """
+    From customer location, search for nearest c-store.
+    Returns:
+        str: Spoken address of nearest c-store.
+        float: distance to it in miles
+    """
+    customer_position = get_customer_location()
+
+    # Do the search for nearby C Store
+    try:
+        response = location.search_place_index_for_text(IndexName=LOCATION_PLACE_INDEX_NAME,
+                                                        Text="Convenience Store",
+                                                        BiasPosition=customer_position)
+
+        # Grab address and make it sound nice - could be a lot more sophisticated here
+        address = response['Results'][0]['Place']
+        spoken_address = get_nice_address(address)
+        store_position = address['Geometry']['Point']
+
+        # How far away is that?
+        shop_dist_km = distance_km(store_position, customer_position)
+        shop_dist_miles = 0.6214 * shop_dist_km
+
+        address_dump = json.dumps(address, default=str)
+
+        logger.info(f"Closest convenience store to {customer_position} is at {store_position}, "
+                    f"with spoken address {spoken_address} and distance {shop_dist_km:0.0f}km"
+                    f" ({shop_dist_miles:0.0f} miles). Full address data: {address_dump}")
+
+    except Exception as e:
+
+        logger.error('Cannot do place index search - perhaps your Location resources were not deployed?'
+                     f'Message: {e} LOCATION_PLACE_INDEX_NAME: {LOCATION_PLACE_INDEX_NAME}')
+
+        spoken_address = "640 Elk Street"
+        shop_dist_miles = 3
+
+    return spoken_address, shop_dist_miles
+
+
+def get_basket_id(handler_input):
+    """
+    We saved our basket ID into Alexa session parameters. Retrieve it.
+    Args:
+        handler_input: As passed into the Alexa skill Lambda handler.
+
+    Returns:
+        str: Basket ID
+    """
+    session_attr = handler_input.attributes_manager.session_attributes
+    if 'BasketId' in session_attr:
+        return session_attr['BasketId']
+    else:
+        return None
```

Comment:
Is there a reason why the Carts service is not being used here?

Reply:
We have talked in our team and it is not clear. We cannot see how using the native carts service would not be fine to do also. Possibly it was simply considered expedient during development to keep the complexity down. I am hoping it is not a blocker for the current iteration?

Reply:
Yes, I think this is a problem. This project is primarily an educational tool for customers to learn how to build similar experiences in their applications. Specifically, this breaks the notion of multi-modal user experiences where, say, I start putting items in a cart using the Alexa skill and then I want to complete my purchase using my phone or a kiosk. By having two cart implementations, continuity is lost. It adds confusion to customers as to why these choices were made and will make the project more difficult to maintain in the future.